File: Matlab/elbow/matrices3D.m

function [ddr,ddr2,ddz,ddz2,ddx,ddx2,ddrz,ddzx,ddrx]= matrices3D (nr,nz,nx,dr,dr2,dz,dz2,dx,dx2)


  ddr=kron(speye(nz,nz),dr);
  ddr=kron(speye(nx,nx),ddr);
  %2 r derivatives
  ddr2=kron(speye(nz,nz),dr2);
  ddr2=kron(speye(nx,nx),ddr2);
   %1 z derivatives
  ddz=kron(dz,speye(nr,nr));
  ddz=kron(speye(nx,nx),ddz);
  
  %2 z derivatives
  ddz2=kron(dz2,speye(nr,nr));
  ddz2=kron(speye(nx,nx),ddz2);
  %1 thetaderivatives
  ddx=kron(dx,speye(nz,nz));
  ddx=kron(ddx,speye(nr,nr));
  %2 theta derivatives
  ddx2=kron(dx2,speye(nz,nz));
  ddx2=kron(ddx2,speye(nr,nr));
  %mixing derivatives (r-z)
  ddrz=kron(dz,dr); 
  ddrz=kron(speye(nx,nx),ddrz); 
  %mixing derivatives (z-theta)
  ddzx=kron(dx,dz); 
  ddzx=kron(ddzx,speye(nr,nr)); 
  %mixing derivatives (r-theta)
  ddrx=kron(dx,speye(nz,nz)); 
  ddrx=kron(ddrx,dr);

File: Python/elbow/Matrices3d.py
import scipy.sparse as sp

def matrices3D(nr, nz, nx, dr, dr2, dz, dz2, dx, dx2):
    """
    Computes 3D differentiation sparse matrices using Kronecker products.
    Translated from MATLAB matrices3D.m
    """
    ddr = sp.kron(sp.eye(nx, format='csr'), sp.kron(sp.eye(nz, format='csr'), dr, format='csr'), format='csr')
    ddr2 = sp.kron(sp.eye(nx, format='csr'), sp.kron(sp.eye(nz, format='csr'), dr2, format='csr'), format='csr')
    
    ddz = sp.kron(sp.eye(nx, format='csr'), sp.kron(dz, sp.eye(nr, format='csr'), format='csr'), format='csr')
    ddz2 = sp.kron(sp.eye(nx, format='csr'), sp.kron(dz2, sp.eye(nr, format='csr'), format='csr'), format='csr')
    
    ddx = sp.kron(sp.kron(dx, sp.eye(nz, format='csr'), format='csr'), sp.eye(nr, format='csr'), format='csr')
    ddx2 = sp.kron(sp.kron(dx2, sp.eye(nz, format='csr'), format='csr'), sp.eye(nr, format='csr'), format='csr')
    
   # ddrz = sp.kron(sp.eye(nx, format='csr'), sp.kron(dz, dr, format='csr'), format='csr')
   # ddzx = sp.kron(sp.kron(dx, dz, format='csr'), sp.eye(nr, format='csr'), format='csr')
   # ddrx = sp.kron(sp.kron(dx, sp.eye(nz, format='csr'), format='csr'), dr, format='csr')
    ddrz=ddr*ddz
    ddzx=ddz*ddx
    ddrx=ddr*ddx
    return ddr, ddr2, ddz, ddz2, ddx, ddx2, ddrz, ddzx, ddrx